lib/org/chromium/webidl/Diff.js

```diff
@@ -170,66 +298,82 @@ foam.CLASS({
-        comparisons. Once all members have been compared, a DiffChunk is
-        generated for all unmatched members.`,
+          members of right. If no differences are found between two objects,
+          they are considered to be matched and removed from the remaining
+          comparisons. Once all members have been compared, a DiffChunk is
+          generated for all unmatched members.`,
       name: 'arrayCmp_',
+      args: [
+        {
+          documentation: 'Value of the left side that is diffed.',
+          name: 'left',
+        },
+        {
+          documentation: 'Value of the right side that is diffed.',
+          name: 'right',
+        },
+        {
+          documentation: 'Identification information for the diff chunk.',
+          name: 'id',
+          typeName: 'Identifier',
+        },
+        {
+          documentation: 'The array tracking diff chunks that are produced.',
+          name: 'opt_changes',
+          typeName: 'Array',
+        },
+      ],
       returns: '',
       code: function arrayCmp_(left, right, id, opt_changes) {
         // Items will be removed from the array once matched.
         // Work on clones to prevent altering original data.
         left = foam.Array.clone(left);
         right = foam.Array.clone(right);
 
-        var matched = false;
         for (var i = 0; i < left.length; i++) {
+          var matched = false; // Used to determine if inner loop broke early.
           for (var j = 0; j < right.length; j++) {
-            // Get id field of left and right.
             var leftName = left[i].id;
             var rightName = right[j].id;
 
             // If names differ, they are definitely not the same.
             if (leftName !== rightName) continue;
-            if (!leftName && !rightName) {
-              // These items have no names. Perform a diff with changes
-              // to determine whether these items are the same.
-              var changes = opt_changes || [];
-              var initLen = changes.length;
-
-              var nextId = Object.assign({}, id,
-                  { path: id.path.concat(`[${leftName}]`) });
-              this.diff_(left[i], right[j], nextId, changes);
-
-              // No differences were found between the two objects.
-              if (changes.length === initLen) {
-                matched = true;
-                break;
-              }
-            } else {
-              // Names are the same. Perform a diff of definition.
-              var nextId = nextId = Object.assign({}, id,
-                  { path: id.path.concat(`[${leftName}]`) });
-              this.diff_(left[i], right[j], nextId);
 
+            // Both items have no names, or they have same names at this point.
+            // Perform a diff with 'changes' so the chunks are not immediately
+            // persisted.
+            var changes = [];
+
+            var nextId = id.clone();
+            nextId.leftKey += `[${i}]`;
```

Comment:
Let's not special case these keys. We want `foo.bar.7.baz`, not `foo.bar[7].baz`